### DISCUSSION TITLE
Utility functions for working with bloom filters

### DISCUSSION THREAD

**Discussion by typedarray** [wevm/viem#429] [Idea / Feature Request]
Bloom filters are a core Ethereum data structure, so it seems reasonable for viem to offer utility functions for working with them.

### Use case
My particular use case is checking if a block contains logs matching a log filter (`addresses` and `topics`) using the `block.logsBloom` field. (I actually can't think of any other use cases, because AFAIK this is the only bloom filter that Ethereum exposes externally).

### References
Web3.js has a few [utility functions](https://web3js.readthedocs.io/en/v1.2.4/web3-utils.html) for bloom filters. They appear to be a re-export of the [ethereum-bloom-filters package](https://github.com/joshstevens19/ethereum-bloom-filters), which could be a good starting point.

### API
We could start with just a single function to check if a value is present in a bloom filter. This function would validate that the bloom filter is a valid Ethereum logs bloom filter, and return a boolean.
```ts
export type IsInBloomFilterParameters = {
  value: Hex | ByteArray
  bloom: Hex | ByteArray
}

function isInBloomFilter(parameters: IsInBloomFilterParameters): boolean
```

We could also consider supporting a plural version that accepts multiple values, and returns early if any values match. This would be useful for the use case described above.
```ts
export type IsInBloomFilterParameters = {
  value: Hex | ByteArray | (Hex | ByteArray)[]
  bloom: Hex | ByteArray
}

// ...
```

**Comment by tmm** [comment]
Currently focused on other things, but down for a contribution!

**Reply by typedarray** [reply]
Same 😅. Going to use `ethereum-bloom-filters` for now, but would love to contribute a viem implementation and run some benchmarks at some point. cc @olastenberg who did some work with bloom filters in js/ts recently.

**Comment by joshstevens19** [comment]
Would you be willing to accept a PR with ethereum bloom filter package installed? I am the creator of that package happen to do a PR here so others can use it?

**Reply by jxom** [reply]
Hey @joshstevens19! If we were to ever implement this, would you mind if we adapt your code (with references to your repo of course / add you as coauthor) into Viem? This way, we can still keep our dependency tree fairly slim, and not have to pull in `js-sha3` (we use `@noble/hashes`).

**Comment by kyscott18** [comment]
I'm willing to contribute by adding some bloom filter utilities. I implemented some bloom filter utilities in Ponder (https://github.com/ponder-sh/ponder/blob/main/packages/core/src/sync-realtime/bloom.ts) using viem utilities based off of Alloy.

### API

```ts
const isLogInBloomFilter: (params: { bloomFilter: Hex, log: Pick<Log, "address" | "topics"> }) => boolean;
```

Could also be nice to have `zeroLogsBloom` or `accrueLogsBloom`.

**Reply by jxom** [reply]
Bloom filters are being removed from the protocol: https://github.com/ethereum/EIPs/pull/8368

Unsure if worth supporting.

**Reply by kyscott18** [reply]
I think it would still be useful for:

1) Historical blocks that still contain `logsBloom`

2) Chains that don't implement this EIP

but completely up to you
